### PR TITLE
[Perf][Mamba] Optimize SSD chunk scan with auto-pipeline and micro-block factorization

### PR DESCRIPTION
**Pull request by stelladuyx**
## Summary

This PR optimizes the Mamba-2 SSD chunk scan forward kernel with two key improvements:
1. **Auto-pipeline optimization**: Leveraging TileLang's built-in auto-pipeline for better instruction overlap
2. **Micro-block factorization**: Reducing exponential operations in diagonal blocks via factorized computation

## Motivation

The SSD chunk scan kernel is compute-bound on the diagonal path exponential operations. The original implementation had two bottlenecks:

1. **Manual pipelining limitations**: Hand-written double-buffering didn't use async copies, limiting memory-compute overlap
2. **Redundant exponentials**: Each element in diagonal blocks computed `exp(dA[l] - dA[s])` independently, with ~6% theoretical overhead

### 1. Auto-Pipeline Optimization

**Before**: Manual double-buffer implementation without async memory operations
```python
# Manual pipeline with synchronous loads
for n_blk in range(T.ceildiv(N, block_n)):
    # Load, compute, store in lock-step
```

**After**: TileLang auto-pipeline with optimal staging
```python
for n_blk in T.Pipelined(T.ceildiv(N, block_n), num_stages=3):
    # Auto-managed async loads with optimal overlap
```

**Impact**: 
- Optimal `num_stages=3` determined via benchmarking (tested 1, 3, 5)
- Better overlap of global memory loads with tensor core operations
- Cleaner code with equivalent or better performance

### 2. Micro-block Factorization for Diagonal Path

**Optimization**: For diagonal blocks where `s` and `l` are close, factorize the exponential:

```
exp(dA[l] - dA[s]) = exp(dA[l] - anchor) * exp(anchor - dA[s])
```

where `anchor = dA[l_block_start]` for each 32×32 micro-block.

**Before**: 64×64 direct computation
```python
for ll, ss in T.Parallel(64, 64):  # 4096 exp calls
    exp_factor = T.exp(dA[l] - dA[s])
    lcb_cast[ll, ss] = cb[ll, ss] * exp_factor * dt[ss]
```

**After**: 2×2 micro-block factorization (for block_l=block_s=64)
```python
# Precompute factorized terms (64 + 2*64 = 192 exp calls)
for ll in T.Parallel(64):
    exp_l_micro[ll] = T.exp(dA[ll] - anchor[r])
for r, ss in T.Parallel(2, 64):
    exp_s_micro[r, ss] = T.exp(anchor[r] - dA[ss]) * dt[ss]

# Four 32×32 micro-blocks:
# - Diagonal blocks (mr=0,mc=0 and mr=1,mc=1): direct exp (2*1024 = 2048)
# - Lower block (mr=1,mc=0): factorized (1024, reuses precomputed)
# - Upper block (mr=1,mc=1): zeros (masked out by causality)
```

**Exponential count reduction**:
- Original: 4096 exp operations per 64×64 block
- Optimized: 192 (precompute) + 2048 (diagonal micro-blocks) = 2240 exp operations
- **Reduction: ~45% fewer exp calls in diagonal blocks**

**Implementation note**: Currently uses Python static unrolling for the 2×2 decomposition, limiting applicability to `block_l=block_s=64` configuration. Future work could generalize this with dynamic tiling.

## Numerical Correctness

Micro-block factorization introduces two-step rounding but maintains acceptable numerical accuracy:

**Error Analysis**:
- Mean absolute error: 1.97e-06 (negligible)
- Max absolute error: 7.75e-02 (rare outliers, 0.0003% of elements)
- All outputs pass `torch.allclose(rtol=1e-3, atol=1e-3)` against reference

Error sources:
1. Two-stage rounding: `exp(a-b)` vs `exp(a-c) * exp(c-b)` 
2. Both mathematically equivalent but different floating-point paths
3. Verified no data races; errors are deterministic from FP arithmetic

The error is acceptable for ML workloads and significantly below typical FP16 training noise.

## Testing

- ✅ All existing unit tests pass (`test_ssd_chunk_scan_fwd`)
- ✅ Numerical correctness verified against PyTorch reference
- ✅ Tested on smoke and full test configurations
- ✅ No regressions on non-optimized code paths

## Three-Way Comparison

| Config | TileOPs-OLD (ms) | TileOPs-NEW (ms) | mamba_ssm (ms) | NEW vs OLD | NEW vs mamba_ssm |
|--------|------------------|------------------|----------------|------------|------------------|
| latency-130m-4k | 0.0951 | 0.0930 | 0.1420 | **1.02x faster** ✅ | **1.53x faster** ✅ |
| serving-130m-4k | 0.3991 | 0.3403 | 0.4395 | **1.17x faster** ✅ | **1.29x faster** ✅ |
| longctx-130m-32k | 0.7445 | 0.6456 | 0.7952 | **1.15x faster** ✅ | **1.23x faster** ✅ |
| latency-370m-4k | 0.1692 | 0.1465 | 0.2106 | **1.15x faster** ✅ | **1.44x faster** ✅ |
| serving-370m-4k | 0.9359 | 0.7946 | 1.0044 | **1.18x faster** ✅ | **1.26x faster** ✅ |
| longctx-370m-32k | 1.8240 | 1.5885 | 1.9419 | **1.15x faster** ✅ | **1.22x faster** ✅ |
| throughput-370m-2k | 1.8128 | 1.5905 | 1.9422 | **1.14x faster** ✅ | **1.22x faster** ✅ |
| latency-2p7b-4k | 0.2012 | 0.1755 | 0.2466 | **1.15x faster** ✅ | **1.41x faster** ✅ |
| serving-2p7b-4k | 1.2003 | 1.0108 | 1.2210 | **1.19x faster** ✅ | **1.21x faster** ✅ |

## Future Work

1. **Generalize micro-block size**: Make factorization work for arbitrary block sizes, not just 64×64
2. **Auto-tuning**: Dynamically select micro-block size based on problem dimensions
3. **Larger workloads**: Test on production-scale sequence lengths (4K-32K tokens)

## References

- Original implementation: `ssd_chunk_scan_before_microtile.py`
- Mamba-2 paper: https://arxiv.org/abs/2405.21060
- TileLang auto-pipeline: https://github.com/TileOPs/TileLang

---

**Checklist**:
- [x] Code follows project style guidelines
- [x] Tests pass
- [x] Performance benchmarks included
- [x] Numerical accuracy verified
- [x] Documentation updated

### CODE REVIEW

File: tileops/kernels/mamba/ssd_chunk_scan.py [hunk 1]
```diff
@@ -281,7 +291,7 @@ def main(
                 T.sync_threads()
 
                 # Only iterate over s-blocks that have at least one s <= l_max.
-                for s_blk in T.serial(T.ceildiv(l0 + block_l, block_s)):
+                for s_blk in T.Pipelined(T.ceildiv(l0 + block_l, block_s), num_stages=num_stages):
```
[the remark this answers is not in the document]
Comment (stelladuyx) [reply tileops/kernels/mamba/ssd_chunk_scan.py:294]: Thanks for the review. I do not think the current evidence shows a race in the s_blk pipeline.

In the generated CUDA, `cb_tile` and `x_tile` are stage-buffered with a ring buffer. `exp_s` is consumed into the thread-local `lcb_cast` fragment before the next iteration overwrites it. Also, for` block_l=block_s=64`, the micro-factorization path only runs on the final diagonal` s_blk` iteration, so row_factor and col_factor are not live across a later iteration.

Therefore, I will keep T.Pipelined for now.

File: tileops/kernels/mamba/ssd_chunk_scan.py [hunk 2]
```diff
@@ -341,41 +351,106 @@ def main(
...
+
+                        # Micro-block factorization: only for block_l=block_s=64, M=32
+                        if diagonal_microtile_size == 32 and block_l == 64 and block_s == 64:
+                            # Optimized 2×2 micro-block factorization
+                            # Only compute factors for the strict-lower block (1,0)
+                            M = 32
+
+                            # Allocate minimal shared memory: only M elements each
+                            row_factor = T.alloc_shared((M,), accum_dtype)
+                            col_factor = T.alloc_shared((M,), accum_dtype)
```
[the remark this answers is not in the document]
Comment (stelladuyx) [reply tileops/kernels/mamba/ssd_chunk_scan.py:363]: Thanks for the suggestion. I do not think moving these allocations is necessary here.

`diagonal_microtile_size`, `block_l`, and `block_s` are compile-time specialization parameters, so this is not a runtime shared-memory allocation inside a dynamic branch. The generated CUDA hoists the storage to block-level `extern __shared__ `memory and aliases it with the earlier exp_l / exp_s buffers after their lifetimes end.

Moving the buffers to the top level may make the lifetime more conservative and increase shared-memory usage, so I will keep the current placement unless we observe a TileLang compilation or correctness issue.

**Review by zhen8838 — APPROVED** [review]
Approved. Static review found no blocking issue in the SSD chunk scan kernel update; latest GitHub CI is green, including gpu-smoke. I did not run local tests or benchmarks.
